pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -1193,6 +1246,68 @@ func (nc *Controller) handleDisruption(zoneToNodeConditions map[string][]*v1.Nod
 	}
 }
 
+func (nc *Controller) podUpdated(oldPod, newPod *v1.Pod) {
+	if newPod == nil {
+		// Nothing to do when pod is deleted.
+		return
+	}
+	if len(newPod.Spec.NodeName) != 0 && (oldPod == nil || newPod.Spec.NodeName != oldPod.Spec.NodeName) {
+		nc.podUpdateQueue.Add(newPod)
+	}
+}
+
+func (nc *Controller) doPodProcessingWorker() {
+	for {
+		obj, shutdown := nc.podUpdateQueue.Get()
+		// "podUpdateQueue" will be shutdown when "stopCh" closed;
+		// we do not need to re-check "stopCh" again.
+		if shutdown {
+			return
+		}
+
+		pod := obj.(*v1.Pod)
+		nodeName := pod.Spec.NodeName
+
+		nc.processPod(nodeName, pod)
+	}
+}
+
+func (nc *Controller) processPod(nodeName string, pod *v1.Pod) {
+	defer nc.podUpdateQueue.Done(pod)
+
+	nodeHealth := nc.nodeHealthMap.getDeepCopy(nodeName)
+	if nodeHealth == nil {
+		// Pod will be handled by standard eviction pass,
+		// when node data is collected.
+		return
+	}
+
+	node, err := nc.nodeLister.Get(nodeName)
+	if err != nil {
+		klog.Errorf("Failed to read node %v.", nodeName)
+		return
+	}
+
+	_, currentReadyCondition := nodeutil.GetNodeCondition(nodeHealth.status, v1.NodeReady)
+	if currentReadyCondition == nil {
+		// We can simply return from here. Even if nodes cache is not up-to-date,
+		// the pod will be handled correctly (evicted if needed) during processing of the node update event.
+		return
+	}
+
+	pods := []*v1.Pod{pod}
+	// Taint based eviction is handled by TaintManager.
+	if !nc.useTaintBasedEvictions {
+		nc.processNoTaintBaseEviction(node, currentReadyCondition, nc.nodeMonitorGracePeriod, pods)
+	}
+
+	if currentReadyCondition.Status != v1.ConditionTrue {
+		if err := nodeutil.MarkPodsNotReady(nc.kubeClient, pods, nodeName); err != nil {
+			utilruntime.HandleError(fmt.Errorf("Unable to mark pods NotReady on node %v: %v", nodeName, err))
```

Comment:
don't we need other error handling here?

Reply:
Requeue added.